### ISSUE TITLE
Specify application directory

### ISSUE THREAD

**Issue by HugLifeTiZ** [whalehub/ginstall.sh#2]
Hi there. I think it'd be useful to be able to specify the application directory to install applications into. I'd prefer not to run scripts like this as root.

Perhaps the script can check inside `${XDG_CONFIG_HOME:-$HOME/.config}` for a configuration file and load it in order to define `$APP_DIR`?

**Comment by whalehub** [comment]
@HugLifeTiZ I'm definitely open to implementing this feature, but I'd rather not have it rely on configuration files or environment variables since that would make it less portable and less flexible. I think I'll instead add a new flag `--directory`/`-d` for specifying a custom application directory.

**Comment by whalehub** [comment]
@HugLifeTiZ I just implemented this feature in the newly release [v1.6.0](https://github.com/whalehub/ginstall.sh/releases/tag/v1.6.0). Feel free to give it a try and let me know how it works for you. 👍

**Comment by HugLifeTiZ** [comment]
That's working for me. This usage isn't really ideal though. I just had an idea; what if ginstall sets its default location to the directory it's located in? That should be a sensible default, right?

**Comment by whalehub** [comment]
@HugLifeTiZ I'm afraid the default location will remain `/usr/local/bin` since that allows applications to be available system-wide, but in case this is just about the hassle of having to type out the custom installation path every time, one option would be to create an alias in your `.bashrc` file to make `ginstall.sh` execute `ginstall.sh -d /path/to/dir` instead.

**Comment by HugLifeTiZ** [comment]
I think that sort of thing is for the user to decide, not the developer. If the user wants the applications to be available system-wide, they can put it in `/usr/local/bin`, or a user-writeable directory that is available system-wide. If they only want it for themselves, they can put it in `$HOME/bin` or whatever they have for their user bin directory. Wherever ginstall.sh lives, it should presume that to be the default install directory. The only case I can think of where that wouldn't work would be if you want only one specific user to be able to install apps, but that user would have sudo access anyways and the script wouldn't be able to do anything when invoked by other users if it lived in `/usr/local/bin`.

Regarding an alias, that is not presently feasible because for some reason you've split the directory argument into three separate arguments for three separate switches. Why is there a `--first-run-dir` and `--rm-from-dir` when you can just use `-d/--directory` for all three usecases?

**Comment by whalehub** [comment]
@HugLifeTiZ The reason is this: Applications that are installed in a user's home directory instead of `/usr/local/bin` are only available to that user. They are neither available to other users nor to system users for use in systemd services. That is why both this tool and the package managers used by the major Linux distributions install applications to a directory that is available system-wide by default.

If your specific use case requires applications to only be accessible by a single user, you're free to use the provided command line flag to install them in the user's home directory, but I won't be changing the default location to a non-standard one to cater to a niche use case.

As for the separate flags for specifying a custom directory, I'd definitely prefer having all of them share the same one as well, but when I initially tried to implement things that way, it lead to some unforeseen and potentially dangerous bugs when a user doesn't specify the correct syntax. That's why I opted to go with a less user-friendly but more robust and secure implementation until I can figure out a better solution.

In the meantime, there something you could use that works for all three flags: a function instead of an alias in `.bashrc`. It should look something like this:

```
ginstall() {
  /path/to/ginstall.sh $1 /path/to/custom/dir $2 $3
}
export -f ginstall
```

This would allow you to install and uninstall applications like this:

```
whalehub@pdh:~# ginstall -d traefik latest
√ traefik_v2.2.1_linux_amd64.tar.gz done (sha256 OK)
The latest version of traefik was installed successfully.
whalehub@pdh:~# ls /home/whalehub/apps
traefik
whalehub@pdh:~#
```

```
whalehub@pdh:~# ginstall --rm-from-dir traefik
removed '/home/whalehub/apps/traefik'
traefik was uninstalled successfully.
whalehub@pdh:~#
```

**Comment by whalehub** [comment]
@HugLifeTiZ After a lot of testing and experimenting, I've found a good way to obsolete the `--first-run-dir` and `--rm-from-dir` flags without compromising the security of the script. They will be removed in the next version, which will leave only the `--directory` flag.

If you want, I could also add a separate flag that automatically uses the current directory (maybe `--current-dir`).

**Comment by HugLifeTiZ** [comment]
Very nice work. A singular `--directory` flag is all that is necessary to make an alias for the script that uses a non-`/usr/local/bin` installation directory and make usage consistent. Thank you!

**Comment by whalehub** [comment]
@HugLifeTiZ Just published [v2.0.0](https://github.com/whalehub/ginstall.sh/releases/tag/v2.0.0) which includes the fix for this. My release numbers use semantic versioning, which means that this is a breaking release. Please make sure to read the full changelog and revamped `--help` flag output. 👍
